experimental/oak_async/src/grpc.rs

```diff
@@ -0,0 +1,146 @@
+    #[doc(hidden)]
+    pub fn new(writer: ChannelResponseWriter) -> OneshotWriter<T> {
+        OneshotWriter {
+            writer,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    /// Send the response, automatically closing the channel.
+    pub fn send(self, msg: &T) -> Result<(), OakError> {
+        self.writer.write(msg, WriteMode::Close)?;
+        // Prevents running the destructor
+        core::mem::forget(self);
+        Ok(())
+    }
+
+    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
+    pub fn into_inner(self) -> ChannelResponseWriter {
+        into_inner(self, |s| &mut s.writer)
+    }
+}
+
+impl<T> Drop for OneshotWriter<T> {
+    /// Close the channel with an `Unknown` status.
+    fn drop(&mut self) {
+        // TODO(wildarch) or should we panic (after setting the error)?
+        error!("drop() called on OneshotWriter");
+        if let Err(e) = self
+            .writer
+            .close(Err(oak::grpc::build_status(Code::Unknown, "")))
+        {
+            error!("Failed to close OneshotWriter in Drop: {}", e);
+        };
+    }
+}
+
+/// Writes multiple messages to a gRPC response channel.
+///
+/// This type is used for gRPC methods where the return type is streaming.
+///
+///
+/// In exceptional circumstances the `into_inner` method may be used to access the underlying
+/// `ChannelResponseWriter`.
+pub struct MultiWriter<T> {
+    writer: ChannelResponseWriter,
+    _msg: core::marker::PhantomData<T>,
+}
+
+impl<T: ::prost::Message> MultiWriter<T> {
+    // Only for use by the Oak Service Generator.
+    #[doc(hidden)]
+    pub fn new(writer: ChannelResponseWriter) -> MultiWriter<T> {
+        MultiWriter {
+            writer,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    /// Send a response to the channel.
+    pub fn send(&self, msg: &T) -> Result<(), OakError> {
+        self.writer.write(msg, WriteMode::KeepOpen)
+    }
+
+    /// Closes the channel with the given error code and message.
+    ///
+    /// To return an `Ok` status, prefer simply dropping the writer or explicitly calling `drop()`.
+    pub fn close_error<S: AsRef<str>>(self, code: Code, msg: S) -> Result<(), OakError> {
+        self.writer
+            .close(Err(oak::grpc::build_status(code, msg.as_ref())))?;
+        // Avoid `Drop`, which would try to close with `Ok`.
+        core::mem::forget(self);
+        Ok(())
+    }
+
+    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
+    pub fn into_inner(self) -> ChannelResponseWriter {
+        into_inner(self, |s| &mut s.writer)
+    }
+}
+
+impl<T> Drop for MultiWriter<T> {
+    /// Close the channel with an `Ok` status.
+    fn drop(&mut self) {
+        if let Err(e) = self.writer.close(Ok(())) {
+            error!("Failed to close MultiWriter in Drop: {}", e);
+        };
+    }
+}
+
+// Helper method for extracting the inner writer without calling `Drop` on `self`
```

Comment:
I don't think this helper is needed if both `struct` move to use an `Option<ChannelResponseWriter>`

Reply:
Removed now that we are using an `Option<..>`